### ISSUE TITLE
Refining conditions

### ISSUE THREAD

**Issue by vyasr** [glotzerlab/signac-flow#177]
There are a couple of things I find confusing about the existing conditions that should perhaps be discussed before a 1.0 release. First, should the `never` and `always` conditions be part of `pre` rather than `_condition`? Placing them as post-conditions just flips their meaning, so it's well-defined, but I'm not sure it conceptually makes sense to put them as post-conditions I think it's a worthwhile simplification. Note that the current behavior also allows nonsensical specifications such as tagging an operation with `pre.never` and `post.never`.

Second, even though neither of these take any arguments, I would still prefer to change them to require parenthesis for consistency. I never remember that the expected syntax for these is different than for other conditions. There are pros and cons to making this change, but in my opinion consistency is more important here, especially because I assume that these are among the least-used conditions so users will expect them to match the behavior of other conditions that they have already used. Is that consistent with what others have experienced?

**Comment by vyasr** [comment]
@csadorf would be interested in your input on this since I believe you originally implemented these conditions.

**Comment by csadorf** [comment]
> First, should the `never` and `always` conditions be part of `pre` rather than `_condition`? Placing them as post-conditions just flips their meaning, so it's well-defined, but I'm not sure it conceptually makes sense to put them as post-conditions I think it's a worthwhile simplification.

I agree that some of those are nonsensical, but not in the way that you describe it. Consider this:

- `@pre.always`: Since the preconditions are additive and "no preconditions" is equivalent to `@pre.always`, adding this never has any effect.
- `@pre.never`: This makes a lot of sense, it unequivocally disables the operation.
- `@post.always`: This makes no sense, since any other post-condition that is negative, would mark this operation as eligible to run.
- `@post.never`: This makes sense, since it ensures that the operation's post-conditions are never satisfied and it makes it so that the operation will always run as long as its preconditions are satisfied.

In conclusion, the only conditions that we should keep in my opinion are `@pre.never` and `@post.never`.

> Note that the current behavior also allows nonsensical specifications such as tagging an operation with `@pre.never` and `@post.never`.
> 
Yes, but since the condition functions are independent from each other, I think that's ok. The only thing that I find actually confusing is adding @`pre.never` and `@pre.always` since it is equivalent to just adding `@pre.never`.

> Second, even though neither of these take any arguments, I would still prefer to change them to require parenthesis for consistency. I never remember that the expected syntax for these is different than for other conditions. There are pros and cons to making this change, but in my opinion consistency is more important here, especially because I assume that these are among the least-used conditions so users will expect them to match the behavior of other conditions that they have already used. Is that consistent with what others have experienced?

I don't think that making them callable when they should never expect any kind of argument makes sense and I certainly had never an issue remembering the right API, but I probably also don't count as the average user. So I'd propose to enable both which is a pattern that you see also in other APIs within the Python ecosystem.

**Comment by vyasr** [comment]
I think I mostly agree with you. However, if your logic for not including `pre.always` is that it's equivalent to having no preconditions, I would argue that `post.never` is also redundant because it's equivalent to having no postconditions. In that case, `pre.never` is the only one we should keep (which is also consistent with my practical experience that that's the only one I've ever used).

@bdice @mikemhenry @atravitz @b-butler @zhou-pj I'd appreciate some extra opinions on my final question regarding whether or not the always/never conditions should be callable or not. @csadorf's solution to support both APIs is reasonable, but I'd appreciate some extra input before adding that complexity in case I'm the only one who has this difficulty.

**Comment by bdice** [comment]
@vyasr `post.never` is not redundant if other post-conditions exist. I have used `post.never` to force execution of operations when other post-conditions are met.

I don't think it's important to make these callable. I prefer non-callable.

**Comment by vyasr** [comment]
I guess I assumed that such a use-case would only be for temporary purposes, at which point adding `post.never` seems no less ad hoc than just commenting out the conditions or similar. If it's actually the intended long-term behavior of the workflow, I'm not sure why there would be any post-conditions. Keeping `post.never` seems like it encourages overspecified workflows in that sense. On the other hand, if it's enabling a use-case people like, then that's fine and we should leave it.

**Comment by csadorf** [comment]
I use `@pre.never` and `@post.never` purely for debugging purposes in the exact way that @bdice described it.

However, more importantly, removing pre- or post-conditions (temporarily with comments or permanently doesn't matter) is **not** equivalent to using `@pre.never` or `@post.never` and that is because other operations might be copying those conditions with `@pre.after` or `pre.copy_from` in which case the whole workflow graph would be inadvertently changed.

Consider the following case:
```python
@Project.operation
@Project.post.isfile("foo.txt")
def foo(job):
    pass

@Project.operation
@Project.pre.after(foo)
def bar(job):
    pass
```

If I wanted to temporarily enforce the execution of the *foo*-operation, I could add `@Project.post.never` to its conditions, which would make `bar` no longer eligible to run (that's a good thing), but removing all post-conditions of the foo-operation would make `bar` eligible to run.

In conclusion, I agree with @vyasr that there are some nonsensical condition functions, specifically the *always* conditions. I think we should deprecate those, but leave everything else *as-is*.
